cpp/src/centrality/betweenness_centrality_impl.cuh

```diff
@@ -15,46 +15,382 @@
+                                         handle.get_stream());
+  detail::scalar_fill(handle, distance.data(), distance.size(), invalid_distance);
+  detail::scalar_fill(handle, sigma.data(), sigma.size(), edge_t{0});
+
+  edge_src_property_t<graph_view_t<vertex_t, edge_t, false, multi_gpu>, edge_t> src_sigma(
+    handle, graph_view);
+  edge_dst_property_t<graph_view_t<vertex_t, edge_t, false, multi_gpu>, vertex_t> dst_distance(
+    handle, graph_view);
+
+  auto vertex_partition =
+    vertex_partition_device_view_t<vertex_t, multi_gpu>(graph_view.local_vertex_partition_view());
+
+  if (vertex_frontier.bucket(bucket_idx_cur).size() > 0) {
+    thrust::for_each(
+      handle.get_thrust_policy(),
+      vertex_frontier.bucket(bucket_idx_cur).begin(),
+      vertex_frontier.bucket(bucket_idx_cur).end(),
+      [d_sigma = sigma.begin(), d_distance = distance.begin(), vertex_partition] __device__(
+        auto v) {
+        auto offset        = vertex_partition.local_vertex_partition_offset_from_vertex_nocheck(v);
+        d_distance[offset] = 0;
+        d_sigma[offset]    = 1;
+      });
+  }
+
+  edge_t hop{0};
+
+  while (true) {
+    update_edge_src_property(handle, graph_view, sigma.begin(), src_sigma);
+    update_edge_dst_property(handle, graph_view, distance.begin(), dst_distance);
+
+    auto [new_frontier, new_sigma] =
+      transform_reduce_v_frontier_outgoing_e_by_dst(handle,
+                                                    graph_view,
+                                                    vertex_frontier.bucket(bucket_idx_cur),
+                                                    src_sigma.view(),
+                                                    dst_distance.view(),
+                                                    cugraph::edge_dummy_property_t{}.view(),
+                                                    brandes_e_op_t<vertex_t>{},
+                                                    reduce_op::plus<vertex_t>());
+
+    update_v_frontier(handle,
+                      graph_view,
+                      std::move(new_frontier),
+                      std::move(new_sigma),
+                      vertex_frontier,
+                      std::vector<size_t>{bucket_idx_next},
+                      thrust::make_zip_iterator(distance.begin(), sigma.begin()),
+                      thrust::make_zip_iterator(distance.begin(), sigma.begin()),
+                      [hop] __device__(auto v, auto old_values, auto v_sigma) {
+                        return thrust::make_tuple(
+                          thrust::make_optional(bucket_idx_next),
+                          thrust::make_optional(thrust::make_tuple(hop + 1, v_sigma)));
+                      });
+
+    vertex_frontier.bucket(bucket_idx_cur).clear();
+    vertex_frontier.bucket(bucket_idx_cur).shrink_to_fit();
+    vertex_frontier.swap_buckets(bucket_idx_cur, bucket_idx_next);
+    if (vertex_frontier.bucket(bucket_idx_cur).aggregate_size() == 0) { break; }
```

Comment:
There should be no risk of an infinite loop (unless someone introduces a bug).

- brandes_e_op_t filters things so that only vertices that have never been visited (dst_distance == invalid_distance_) can be considered for addition to the frontier

- update_v_frontier sets the distance flag to hop + 1 on any vertex that was found and adds it to the frontier.

Suppose we have n vertices.  All vertices are initially set to invalid_distance_.  Then the seed vertex is set to distance 0.  The number of unvisited vertices is now n - 1.  Each iteration of the loop will result in two possibilities:

- No vertices of distance invalid_distance are encountered, in which case we exit the loop, or
- k vertices (where k > 1) are discovered and added to the frontier, which will reduce the number of unvisited vertices by k (which is at least 1).

The maximum number of iterations of this loop would be n-1 assuming our graph is a line.  No configuration of parameters would vary this logic sequence.